npm install should run with openattic user (non root user)

ricardoasmarques opened the pull request:
This will prevent the error "EPERM: operation not permitted, chown '/srv/openattic/webui/node_modules/grunt-contrib-copy/package.json'" when running npm install

Signed-off-by: Ricardo Marques <rimarques@suse.com>

rjfd commented:
@ricardoasmarques can you post the contents of `/etc/exports` generated by Vagrant?

ricardoasmarques commented:
@rjfd the content of `/etc/exports`:

```
# VAGRANT-BEGIN: 1000 f727a29f-1582-4ef7-8e5d-3ec1adad831e
"/home/rimarques-local/projects/forks/openattic-rdias" 192.168.121.164(rw,no_subtree_check,all_squash,insecure,anonuid=1000,anongid=100,fsid=3635135649)
"/home/rimarques-local/projects/forks/DeepSea-rdias" 192.168.121.164(rw,no_subtree_check,all_squash,insecure,anonuid=1000,anongid=100,fsid=4121414547)
# VAGRANT-END: 1000 f727a29f-1582-4ef7-8e5d-3ec1adad831e
```

And here is a snippet that reproduces the problem:

Error:
```
salt:/srv/openattic/webui # rm -r node_modules/
salt:/srv/openattic/webui # npm i
npm WARN package.json openattic-web@2.0.0 No README data
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-cssmin/0.14.0/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-ng-annotate/1.0.1/package.tgz
npm ERR! tar.unpack untar error /root/.npm/matchdep/1.0.1/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-clean/1.0.0/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt/0.4.5/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-copy/0.8.2/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-uglify/0.11.1/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-jshint/1.0.0/package.tgz
npm ERR! tar.unpack untar error /root/.npm/load-grunt-tasks/3.4.1/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-jscs/2.7.0/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-rev/0.1.0/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-usemin/3.1.1/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-watch/0.6.1/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-html-build/0.5.3/package.tgz
npm ERR! tar.unpack untar error /root/.npm/grunt-contrib-concat/1.0.0/package.tgz
npm ERR! Linux 4.1.15-8-default
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "i"
npm ERR! node v4.6.1
npm ERR! npm  v2.15.9
npm ERR! path /srv/openattic/webui/node_modules/grunt-contrib-cssmin/package.json
npm ERR! code EPERM
npm ERR! errno -1
npm ERR! syscall chown

npm ERR! Error: EPERM: operation not permitted, chown '/srv/openattic/webui/node_modules/grunt-contrib-cssmin/package.json'
npm ERR!     at Error (native)
npm ERR!  { [Error: EPERM: operation not permitted, chown '/srv/openattic/webui/node_modules/grunt-contrib-cssmin/package.json']
npm ERR!   errno: -1,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'chown',
npm ERR!   path: '/srv/openattic/webui/node_modules/grunt-contrib-cssmin/package.json',
npm ERR!   fstream_finish_call: 'chown',
npm ERR!   fstream_type: 'File',
npm ERR!   fstream_path: '/srv/openattic/webui/node_modules/grunt-contrib-cssmin/package.json',
npm ERR!   fstream_class: 'FileWriter',
npm ERR!   fstream_stack: 
npm ERR!    [ '/usr/lib64/node_modules/npm/node_modules/fstream/lib/writer.js:306:21',
npm ERR!      '/usr/lib64/node_modules/npm/node_modules/graceful-fs/polyfills.js:210:7',
npm ERR!      'FSReqWrap.oncomplete (fs.js:82:15)' ] }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     /srv/openattic/webui/npm-debug.log
```
Success
```
salt:/srv/openattic/webui # rm -r node_modules/
salt:/srv/openattic/webui # runuser -l openattic -c 'cd /srv/openattic/webui && npm i'
npm WARN package.json openattic-web@2.0.0 No README data
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated graceful-fs@1.2.3: graceful-fs v3.0.0 and before will fail on node releases >= v7.0. Please update to graceful-fs@^4.0.0 as soon as possible. Use 'npm ls graceful-fs' to find it in the tree.
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
grunt-rev@0.1.0 node_modules/grunt-rev

grunt-contrib-copy@0.8.2 node_modules/grunt-contrib-copy
├── file-sync-cmp@0.1.1
└── chalk@1.1.3 (escape-string-regexp@1.0.5, ansi-styles@2.2.1, supports-color@2.0.0, strip-ansi@3.0.1, has-ansi@2.0.0)

grunt-contrib-concat@1.0.0 node_modules/grunt-contrib-concat
├── source-map@0.5.6
└── chalk@1.1.3 (escape-string-regexp@1.0.5, ansi-styles@2.2.1, supports-color@2.0.0, has-ansi@2.0.0, strip-ansi@3.0.1)

load-grunt-tasks@3.4.1 node_modules/load-grunt-tasks
├── arrify@1.0.1
├── resolve-pkg@0.1.0 (resolve-from@2.0.0)
├── pkg-up@1.0.0 (find-up@1.1.2)
└── multimatch@2.1.0 (array-differ@1.0.0, array-union@1.0.2, minimatch@3.0.3)

grunt-contrib-clean@1.0.0 node_modules/grunt-contrib-clean
├── async@1.5.2
└── rimraf@2.6.1 (glob@7.1.1)

grunt-contrib-watch@0.6.1 node_modules/grunt-contrib-watch
├── async@0.2.10
├── lodash@2.4.2
├── tiny-lr-fork@0.0.5 (debug@0.7.4, qs@0.5.6, faye-websocket@0.4.4, noptify@0.0.3)
└── gaze@0.5.2 (globule@0.1.0)

matchdep@1.0.1 node_modules/matchdep
├── stack-trace@0.0.9
├── findup-sync@0.3.0 (glob@5.0.15)
├── resolve@1.1.7
└── micromatch@2.3.11 (filename-regex@2.0.0, normalize-path@2.0.1, is-extglob@1.0.0, array-unique@0.2.1, is-glob@2.0.1, extglob@0.3.2, arr-diff@2.0.0, expand-brackets@0.1.5, regex-cache@0.4.3, kind-of@3.1.0, object.omit@2.0.1, parse-glob@3.0.4, braces@1.8.5)

grunt@0.4.5 node_modules/grunt
├── eventemitter2@0.4.14
├── which@1.0.9
├── dateformat@1.0.2-1.2.3
├── async@0.1.22
├── colors@0.6.2
├── getobject@0.1.0
├── lodash@0.9.2
├── rimraf@2.2.8
├── hooker@0.2.3
├── grunt-legacy-util@0.2.0
├── exit@0.1.2
├── nopt@1.0.10 (abbrev@1.1.0)
├── coffee-script@1.3.3
├── iconv-lite@0.2.11
├── minimatch@0.2.14 (sigmund@1.0.1, lru-cache@2.7.3)
├── underscore.string@2.2.1
├── glob@3.1.21 (inherits@1.0.2, graceful-fs@1.2.3)
├── grunt-legacy-log@0.1.3 (grunt-legacy-log-utils@0.1.1, lodash@2.4.2, underscore.string@2.3.3)
├── findup-sync@0.1.3 (lodash@2.4.2, glob@3.2.11)
└── js-yaml@2.0.5 (esprima@1.0.4, argparse@0.1.16)

grunt-contrib-cssmin@0.14.0 node_modules/grunt-contrib-cssmin
├── chalk@1.1.3 (escape-string-regexp@1.0.5, supports-color@2.0.0, ansi-styles@2.2.1, has-ansi@2.0.0, strip-ansi@3.0.1)
├── clean-css@3.4.25 (commander@2.8.1, source-map@0.4.4)
└── maxmin@1.1.0 (figures@1.7.0, gzip-size@1.0.0, pretty-bytes@1.0.4)

grunt-ng-annotate@1.0.1 node_modules/grunt-ng-annotate
├── lodash.clonedeep@3.0.2 (lodash._bindcallback@3.0.1, lodash._baseclone@3.3.0)
└── ng-annotate@1.2.1 (tryor@0.1.2, stringmap@0.2.2, stringset@0.2.1, simple-is@0.2.0, alter@0.2.0, simple-fmt@0.1.0, stable@0.1.5, convert-source-map@1.1.3, ordered-ast-traverse@1.1.1, source-map@0.5.6, acorn@2.6.4, optimist@0.6.1)

grunt-usemin@3.1.1 node_modules/grunt-usemin
├── path-exists@1.0.0
├── chalk@1.1.3 (escape-string-regexp@1.0.5, ansi-styles@2.2.1, supports-color@2.0.0, has-ansi@2.0.0, strip-ansi@3.0.1)
├── debug@2.6.1 (ms@0.7.2)
└── lodash@3.10.1

grunt-html-build@0.5.3 node_modules/grunt-html-build
├── js-beautify@1.6.11 (nopt@3.0.6, config-chain@1.1.11, mkdirp@0.5.1, editorconfig@0.13.2)
└── lodash@3.10.1

grunt-contrib-jshint@1.0.0 node_modules/grunt-contrib-jshint
├── hooker@0.2.3
├── chalk@1.1.3 (escape-string-regexp@1.0.5, supports-color@2.0.0, ansi-styles@2.2.1, has-ansi@2.0.0, strip-ansi@3.0.1)
└── jshint@2.9.4 (strip-json-comments@1.0.4, exit@0.1.2, console-browserify@1.1.0, minimatch@3.0.3, shelljs@0.3.0, cli@1.0.1, htmlparser2@3.8.3, lodash@3.7.0)

grunt-contrib-uglify@0.11.1 node_modules/grunt-contrib-uglify
├── uri-path@1.0.0
├── chalk@1.1.3 (supports-color@2.0.0, escape-string-regexp@1.0.5, ansi-styles@2.2.1, has-ansi@2.0.0, strip-ansi@3.0.1)
├── maxmin@2.1.0 (figures@1.7.0, pretty-bytes@3.0.1, gzip-size@3.0.0)
├── uglify-js@2.6.4 (async@0.2.10, uglify-to-browserify@1.0.2, source-map@0.5.6, yargs@3.10.0)
└── lodash@4.17.4

grunt-jscs@2.7.0 node_modules/grunt-jscs
├── vow@0.4.15
├── hooker@0.2.3
├── lodash@3.10.1
└── jscs@2.9.0 (to-double-quotes@2.0.0, to-single-quotes@2.0.1, reserved-words@0.1.1, natural-compare@1.2.2, estraverse@4.2.0, strip-json-comments@1.0.4, pathval@0.1.1, jscs-preset-wikimedia@1.0.0, esprima@2.7.3, strip-bom@2.0.0, commander@2.9.0, exit@0.1.2, xmlbuilder@3.1.0, chalk@1.1.3, glob@5.0.15, minimatch@3.0.3, cli-table@0.3.1, vow-fs@0.3.6, resolve@1.3.2, jscs-jsdoc@1.3.2, js-yaml@3.4.6, htmlparser2@3.8.3, prompt@0.2.14, jsonlint@1.6.2, escope@3.6.0, babel-jscs@2.0.5)

```

rjfd commented:
@ricardoasmarques apparently the problem is caused by `npm install` when running as root on top of an NFS mount directory. `npm install` tries to change the ownership of some directory/file and that is not allowed by NFS. In order to not break the environments where openATTIC source is not in an NFS mount, we should detect if the `chown` command is allowed.

What do you think of the following code as an alternative to your patch?

```
cd /srv/openattic/webui
touch a
chown openattic a &> /dev/null
rm a
CHOWN_RET=$?
if [ "$CHOWN_RET" == "0" ]; then
  npm install
  bower install --allow-root
else
  runuser -l openattic -c 'cd /srv/openattic/webui && npm install && \
                                       bower install --allow-root'
fi
grunt dev
```

ricardoasmarques commented:
@rjfd thank you for your suggestion, I've updated this PR accordingly, with two small changes:

- removed '--allow-root' on bower install, when running with openattic user
- renamed temporary file 'a' to '.chown'

rjfd commented:
@ricardoasmarques great! can you also update the ubuntu docker image?

ricardoasmarques commented:
@rjfd done for ubunto docker image.
